### PR TITLE
weak_ptr with tests added

### PR DESCRIPTION
**Pull request by jakjan95**
Please check code, lock function need constructor in shared_ptr which take weak_ptr so I left suggestive comment in function lock :)

### CODE REVIEW

File: homework/shared_ptr/weak_ptr.hpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    if (controlBlock_) {
+        controlBlock_->incrementWeakRefs();
+    }
+}
+
+template <typename T>
+weak_ptr<T>::weak_ptr(const cs::shared_ptr<T>& r) noexcept
+    : ptr_{r.ptr_}, controlBlock_{r.controlBlock_} {
+    controlBlock_->incrementWeakRefs();
+}
```
Comment (arafalski) [inline homework/shared_ptr/weak_ptr.hpp]: I think that you should check if `controlBlock_` is not `nullptr`

File: homework/shared_ptr/weak_ptr.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+template <typename T>
+weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr<T>& r) noexcept {
+    ptr_ = r.ptr_;
+    if (controlBlock_) {
+        controlBlock_->incrementWeakRefs();
+    }
+    return *this;
+}
+
```
Comment (arafalski) [inline homework/shared_ptr/weak_ptr.hpp]: You should add controlBlock from the original pointer to new one

File: homework/shared_ptr/weak_ptr.hpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+template <typename T>
+weak_ptr<T>& weak_ptr<T>::operator=(weak_ptr<T>&& r) noexcept {
+    if (this != &r) {
+        delete ptr_;
```
Comment (arafalski) [inline homework/shared_ptr/weak_ptr.hpp]: What if there is another owner of `ptr_`?

File: homework/shared_ptr/weak_ptr.hpp (hunk 2 above)
Comment (Vicx95) [reply homework/shared_ptr/weak_ptr.hpp]: You should check that we don't make self-assignment

**Comment by arafalski** [comment]
Please link issues that will be solved by that PR

**Comment by jakjan95** [comment]
Now everything should be good, I corrected code :)

**Review by arafalski — APPROVED** [review]

**Review by Vicx95 — APPROVED** [review]

**Pull request merged**